recursion/gnark-ffi/src/groth16.rs

```diff
@@ -254,6 +262,108 @@ impl Groth16Prover {
+        let gnark_dir = manifest_dir.join("../gnark");
+        let cwd = std::env::current_dir().unwrap();
+        let data_dir = cwd.join(build_dir);
+        let data_dir_str = data_dir.to_str().unwrap();
+
+        // Run `make`.
+        let make = Command::new("make")
+            .current_dir(&gnark_dir)
+            .stderr(Stdio::inherit())
+            .stdout(Stdio::inherit())
+            .stdin(Stdio::inherit())
+            .output()
+            .unwrap();
+        if !make.status.success() {
+            panic!("failed to run make");
+        }
+
+        // Run the verify script.
+        let result = Command::new("go")
+            .args([
+                "run",
+                "main.go",
+                "verify-groth16",
+                "--data",
+                data_dir_str,
+                "--encoded-proof",
+                &proof.encoded_proof,
+                "--vkey-hash",
+                &proof.public_inputs[0],
+                "--commited-values-digest",
+                &proof.public_inputs[1],
+            ])
+            .current_dir(gnark_dir)
+            .stderr(Stdio::inherit())
+            .stdout(Stdio::inherit())
+            .stdin(Stdio::inherit())
+            .output()
+            .unwrap();
+        if !result.status.success() {
+            panic!("failed to run build script");
+        }
+
+        true
+    }
+
+    /// Generates a Groth16 proof by sending a request to the Gnark server.
+    pub fn convert(proof: Groth16Proof, build_dir: PathBuf) -> SolidityGroth16Proof {
+        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let gnark_dir = manifest_dir.join("../gnark");
+        let cwd = std::env::current_dir().unwrap();
+        let data_dir = cwd.join(build_dir);
+        let data_dir_str = data_dir.to_str().unwrap();
+
+        // Run `make`.
+        let make = Command::new("make")
+            .current_dir(&gnark_dir)
+            .stderr(Stdio::inherit())
+            .stdout(Stdio::inherit())
+            .stdin(Stdio::inherit())
+            .output()
+            .unwrap();
+        if !make.status.success() {
+            panic!("failed to run make");
+        }
+
+        // Run the verify script.
+        let result = Command::new("go")
+            .args([
+                "run",
+                "main.go",
+                "convert-groth16",
+                "--data",
+                data_dir_str,
+                "--encoded-proof",
+                &proof.encoded_proof,
+                "--vkey-hash",
+                &proof.public_inputs[0],
+                "--commited-values-digest",
+                &proof.public_inputs[1],
+            ])
+            .current_dir(gnark_dir)
+            .stderr(Stdio::inherit())
+            .stdout(Stdio::inherit())
+            .stdin(Stdio::inherit())
+            .output()
+            .unwrap();
+        if !result.status.success() {
+            panic!("failed to run build script");
```

Comment:
Will update

Reply:
fyi go best practice errors (at least in Go) generally are just like `"running build script"` since that makes that full bubbled up error message, since the "failed to" or "error:" prefixing is redundant. I think this file had some other places were we do it wrong but I can clean those up in a different PR.